pandas/core/internals.py

```diff
@@ -5440,6 +5449,14 @@ def is_uniform_join_units(join_units):
         len(join_units) > 1)
 
 
+def is_uniform_reindex(join_units):
+    return (
+        # TODO: should this be ju.block._can_hold_na?
+        all(ju.block and ju.block.is_extension for ju in join_units) and
+        len(set(ju.block.dtype.name for ju in join_units)) == 1
```

Comment:
Did you check this for the Categorical case? (I think I asked about it before) It's fine to disallow the categorical case in this check for now (we still need to fix Categorical.take in general in a follow-up PR anyhow), but I think categoricals now pass this check? 
(and now with the current code, they would return object dtype I think)